lib/msgraph/client.go

```diff
@@ -0,0 +1,334 @@
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package msgraph
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/defaults"
+)
+
+// baseURL is the default value for [client.baseURL]. It is the address of MS Graph API v1.0.
+const baseURL = "https://graph.microsoft.com/v1.0"
+
+// defaultPageSize is the page size used when [Config.PageSize] is not specified.
+const defaultPageSize = 500
+
+// scopes defines OAuth scopes the client authenticates for.
+var scopes = []string{"https://graph.microsoft.com/.default"}
+
+type azureTokenProvider interface {
+	GetToken(ctx context.Context, opts policy.TokenRequestOptions) (azcore.AccessToken, error)
+}
+
+// Config defines configuration options for [client].
+type Config struct {
+	// TokenProvider provides tokens to authorize to MS Graph API.
+	// Concrete implementations of this are defined by [github.com/Azure/azure-sdk-for-go/sdk/azidentity].
+	TokenProvider azureTokenProvider
+	// HTTPClient is the HTTP client to use for calls to the API.
+	// If not specified, [http.DefaultClient] is used.
+	HTTPClient *http.Client
+	// Clock is the clock to use for time operations (e.g. delay when retrying requests).
+	Clock clockwork.Clock
+	// RetryConfig specifies parameters for retrying failed requests.
+	// Client will prefer to use the `Retry-After` header returned from the API,
+	// and only use this retry config if the header is not provided.
+	RetryConfig *retryutils.RetryV2Config
+	// PageSize limits the number of objects to return in one batch when using paginated requests (via the `$top` parameter).
+	PageSize int
+}
+
+// SetDefaults sets the default values for optional fields.
+func (cfg *Config) SetDefaults() {
+	if cfg.HTTPClient == nil {
+		cfg.HTTPClient = http.DefaultClient
+	}
+	if cfg.Clock == nil {
+		cfg.Clock = clockwork.NewRealClock()
+	}
+	if cfg.RetryConfig == nil {
+		cfg.RetryConfig = &retryutils.RetryV2Config{
+			First:  1 * time.Second,
+			Driver: retryutils.NewExponentialDriver(1 * time.Second),
+			Max:    defaults.HighResPollingPeriod,
+		}
+	}
```

Comment:
You can simplify a lot of this with `cmp.Or`
```suggestion
	cfg.HTTPClient = cmp.Or(cfg.HTTPClient, http.DefaultClient)
	cfg.Clock = cmp.Or(cfg.Clock, clockwork.NewRealClock())
	// ...
```